### ISSUE TITLE
styled-component does not inherit props passed through attrs method in react-native

### ISSUE THREAD

**Issue by oltioff** [styled-components/styled-components#4183]
## System:
 - OS: macOS 13.5.2
 - CPU: (10) arm64 Apple M2 Pro
 - Memory: 266.34 MB / 16.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 18.18.0 - ~/.nvm/versions/node/v18.18.0/bin/node
 - Yarn: 1.22.19 - ~/Desktop/Developer/mobile-app/node_modules/.bin/yarn
 - npm: 9.8.1 - ~/.nvm/versions/node/v18.18.0/bin/npm
 - Watchman: 2023.09.04.00 - /opt/homebrew/bin/watchman
## npmPackages:
 - babel-plugin-styled-components: ^2.1.4 => 2.1.4
 - styled-components: 6.0.7 => 6.0.7

## Environment
"dependencies": {
  ...,
  "react": "18.2.0",
  "react-is": "18.2.0",
  "react-native": "0.72.5",
  "styled-components": "6.0.7",
  "stylis": "^4.0.0"
},
"devDependencies": {
  ...,
  "@types/react": "^18.0.24",
  "@typescript-eslint/eslint-plugin": "^5.54.1",
  "@typescript-eslint/parser": "^5.54.1",
  "@typescript-eslint/typescript-estree": "^5.54.1",
  "babel-plugin-styled-components": "^2.1.4",
},
resolutions: {
  "styled-components": "^6"
}

## Reproduction

given the following component, typescript recognise *source* and *resizeMode* as valid props

`
const BlurImage = styled.Image.attrs({
  source: blur,
  resizeMode: 'stretch'
})'
  width: 100%;
  height: 100%;
';
`

## Expected Behavior

Typescript doesn't show any error

## Actual Behavior

![image](https://github.com/styled-components/styled-components/assets/144789343/aa2efcc9-72f6-4078-ad9c-1e585fa5ca93)

**Comment by Martinocom** [comment]
I have the same problem.
To solve it I downgraded to `6.0.7` and I typed as _optional_ my default properties.

```jsx
const BlurImage = styled(Image).attrs<{ source?: ImageSourcePropType | undefined }>({
  source: blur,
  resizeMode: 'stretch'
})`
  width: 100%;
  height: 100%;
`;

...

{/* This works now */}
<BlurImage />
```

**This is not a solution** and I would not suggest to use it in production, since `source` becomes optional also for the `attrs` method. If you have any other props, it becomes quickly a very long typing experience. You could use also `.attrs<Partial<ImageProps>>` but it's even more dangerous, since it hides **all** mandatory attributes.

**Comment by winterbe** [comment]
I have the same problem when I try to upgrade from `6.0.7` to `6.1.0`. It's a general TypeScript problem, not specific to React Native.

Here's a simple test to reproduce the issue:

```tsx
function Example(props: {name: string}) {
    return <span>{props.name}</span>
}

const HotExample = styled(Example).attrs({name: "Hot"})`
    color: hotpink;
`

function MyComponent() {
    return <HotExample/>    /* <-- TS Error */
}
```

Corresponding TypeScript `5.2.2` compiler error:

```
TS2769: No overload matches this call.
Overload  1  of  3 ,
(props: PolymorphicComponentProps<"web", FastOmit<FastOmit<Substitute<{ name: string; }, { name: string; }>, never>, never>, void, void, {}, {}>): Element
, gave the following error.
Property  name  is missing in type  {}  but required in type
FastOmit<Substitute<FastOmit<FastOmit<Substitute<{     name: string; }, {     name: string; }>, never>, never>, FastOmit<{}, never>>, keyof ExecutionProps>
Overload  2  of  3 ,
(props: FastOmit<FastOmit<Substitute<{ name: string; }, { name: string; }>, never>, never>): ReactNode
, gave the following error.
Property  name  is missing in type  {}  but required in type
FastOmit<FastOmit<Substitute<{     name: string; }, {     name: string; }>, never>, never>
Overload  3  of  3 ,  (props: { name: string; }): Element , gave the following error.
Property  name  is missing in type  {}  but required in type  { name: string; } 
```

**Comment by tomfridental1** [comment]
This is duplication of: 
https://github.com/styled-components/styled-components/issues/4076
